### PR TITLE
Fix #1010, fix #986: bookmark sync update fixes.

### LINKED ISSUE #986: Bookmark sync changes are not shown if bookmark hanger is open
### Description: 
Bookmark sync changes are not shown if bookmark hanger is open

### Steps to Reproduce 
  1. Create sync chain between two devices (iOS-iOS)
  2. Ensure bookmarks sync between devices
  3. Keep bookmark hanger open in one device and move bookmarks/edit bookmark name and reorder on the other device
  4. Changes are sync'd but not shown on the device where bookmark hanger is open
  5. Close and reopen bookmark tray, all changes are shown 

**Actual result:** 
Bookmark sync changes are not shown if bookmark hanger is open

**Expected result:**
Should show changes without having to close bookmark tray

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Fairly easy 

**Brave Version:** 
1.9 

**Device details:** 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)? 

### Additional Information 
https://github.com/brave/brave-ios/issues/772#issuecomment-460928574

### LINKED ISSUE #1010: brave crashes when attempting to sync into a folder that's opened
### Description:

When you're attempting to move a folder from another device into a folder that iOS has opened, the browser will crash once a sync occurs. It will result in the following crash:

* Crash Stack from Xcode --> [stack.txt](https://github.com/brave/brave-ios/files/3018505/stack.txt)

Screen Shot 2019-03-28 at 2 59 44 AM

### Steps to Reproduce 

@btlechowski thank you for narrowing down the STR and proving the following STR:

1. Clean install on desktop
2. import the following bookmarks under desktop [randomBookmarks.txt](https://github.com/brave/brave-ios/files/3016588/randomBookmarks.txt) <-- rename to `.html`
3. Create a new sync chain on desktop once the above bookmarks have been imported
4. Clean install Brave on iPhone/iPad
5. Join the sync chain that you created on the desktop with the iOS device
6. Wait for sync to complete on both devices
7. On iPhone: Open `Sports` folder in bookmarks and keep it open (prevent the phone from locking etc)
8. On Desktop: Move `Games` bookmark folder to `Sports`
9. On iPhone: Wait for sync while keeping `Sports` folder open

**Note: The key is to keep bookmark folder open on iOS when syncing**

**Actual result:** 

* Crashes with the following Stack from Xcode --> [stack.txt](https://github.com/brave/brave-ios/files/3018509/stack.txt)

**Expected result:**

Application shouldn't be crashing

**Reproduces how often:** [Easily reproduced, Intermittent Issue]

100% reproducible when using the above STR

**Brave Version:** 

* `1.9 (19.03.27.19)`
* `1.9 (1.2.3.4)` (Xcode from source)

```
Kamils-MBP:brave-ios kjozwiak$ git log -1
commit e9fe96e7c4ffd81236b41e302552609a87270ac0 (HEAD -> development, tag: v1.9, origin/development, origin/HEAD)
Merge: 829037e0 b6dac8bc
Author: Michał Buczek <michal.buczek@protonmail.com>
Date:   Wed Mar 27 20:15:04 2019 +0100

    Merge pull request #1009 from brave/update_beta_icons

    No Bug: Update beta icons
```

**Device details:** 

* `iPad Mini 4` running `iOS 12.1.4`
* `iPhone XR` running `12.1 (16B91)` using Xcode emulator

**Website problems only:**
- did you check with Brave Shields down? N/A
- did you check in Safari/Firefox (WkWebView-based browsers)? N/A

### Additional Information

@btlechowski & @LaurenWags have also reproduced this on their devices.

### PR DESCRIPTION
**Pull request by iccub**
## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

File: Data/models/Bookmark.swift [hunk 1]
```diff
@@ -507,10 +507,10 @@ extension Bookmark {
     func updateResolvedRecord(_ record: SyncRecord?, context: WriteContext = .new) {
         guard let bookmark = record as? SyncBookmark, let site = bookmark.site else { return }
         title = site.title
+        syncParentUUID = bookmark.parentFolderObjectId
```
Comment (danishjafri88) [inline Data/models/Bookmark.swift:510]: How does this affect code? I am sorry if I've missed something. trying to figure out

File: Client/Frontend/Menu/BookmarksViewController.swift [hunk 2]
```diff
@@ -188,6 +194,14 @@ class BookmarksViewController: SiteTableViewController {
     
     reloadData()
   }
+    
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        // Make sure to remove fetch results controller when view disappears.
+        // Otherwise, it may result in crash if a user is in a nested folder and
+        // sync changes happen.
+        bookmarksFRC = nil
```
Comment (jhreis) [inline Client/Frontend/Menu/BookmarksViewController.swift:203]: Curious, why is this causing an actual crash when UI not visible? Is the table deallocated and we're trying to update it or something?

File: Data/models/Bookmark.swift (hunk 1 above)
Comment (iccub) [reply Data/models/Bookmark.swift:510]: I just moved it up few lines, previously it was set after the `updateInternal` and it was taking wrong parent Bookmark

File: Client/Frontend/Menu/BookmarksViewController.swift (hunk 2 above)
Comment (iccub) [reply Client/Frontend/Menu/BookmarksViewController.swift:203]: Yeah the FRC callbacks strill trigger even though another view controller was pushed on top of this vc. This causes crash if you move bookmark from root directory to a folder. The `BookmarkVC` of root directory still tries to configure UI.

**Review by jhreis — APPROVED** [review]

**Pull request merged**